### PR TITLE
MPC: Survey: Added a turning radius to altitude changes in FlightModeAuto

### PR DESCRIPTION
**Pull request by Claudio-Chies**
### Solved Problem
When conducting a survey, the drone tended to slow down and accelerate again when covering large altitude changes, with terrain following enabled. 

Example Survey:
image

**Without Fix:**
image
This is best seen when looking at the velocity change in x and z direction in the plot above.

### Solution
To solve this issue i added a turning radius for the vertical component, whereby I simplify the problem so that we consider the 2D case where the vertical component of the incoming vector is the norm of the xy component, and the z component. 
I also removed the requirement for the XY turning radius that the altitude between the current and next waypoint have to be the same. 
**With Fix:**
image
As we can see in the Plot above, the velocity changes close to the waypoints are gone. 
### Changelog Entry
For release notes:
```
Bugfix: removed slowdown during ground following surveys with large altitude changes.
```

### Alternatives
We could also directly generalize it for the 3D case, but that would require alot more changes.

### Test coverage
- Tested in SIH, [Flightreview](https://review.px4.io/plot_app?log=e29141c9-e999-449b-8401-97e31f658ead)
- Tested on Hardware [FlightReview](https://review.px4.io/plot_app?log=3f341aac-88c7-4c30-879e-548dac2dcd83)
image

### CODE REVIEW

**Review by MaEtUgR — APPROVED** [review]
Thanks for looking into it! We also tested on the real vehicle and it solves the issue.
I allowed myself to commit some refactoring which should not change the result.
We should really opt for a 3D solution of this whole calculation because it seems to add duplication make it hard to follow all the steps. But I agree it's a bigger change we need to tackle. I see the vertical calculation is now analogous to https://github.com/PX4/PX4-Autopilot/blob/ab320017ccf5efb6818fbc41d8995b413eb995d4/src/lib/motion_planning/TrajectoryConstraints.hpp#L87-L92 but honestly can't follow and verify every step of both the versions exactly.

File: src/lib/motion_planning/PositionSmoothing.cpp [hunk 1]
```diff
@@ -106,16 +106,33 @@ float PositionSmoothing::_getMaxXYSpeed(const Vector3f(&waypoints)[3]) const
 
 float PositionSmoothing::_getMaxZSpeed(const Vector3f(&waypoints)[3]) const
 {
+	const Vector3f &start_position = waypoints[0];
+	const Vector3f &target = waypoints[1];
+	const Vector3f &next_target = waypoints[2];
 
-	const auto &target = waypoints[1];
+	const Vector2f start_position_xy_z = {start_position.xy().norm(), start_position(2)};
```
[the remark this answers is not in the document]
Comment (Claudio-Chies) [reply src/lib/motion_planning/PositionSmoothing.cpp:116]: good catch, yea youre right. above there is an comment that its done because the position setpoint can been modified by an external source (eg. Obstacle Avoidance)

**Review by MaEtUgR — APPROVED** [review]
Thanks for checking. Your latest commit is exactly what I was thinking about but I wanted to confirm with you 👍

File: src/lib/motion_planning/PositionSmoothing.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Luka-Filipovic) [reply src/lib/motion_planning/PositionSmoothing.cpp:116]: I think there is an issue with the logic here. Calculating `.xy()` for each waypoint means the result can significantly depend on the position of the origin, rather than just the relative positions of the triplet points.

As a simple way to validate my concern, compare the output in two cases: onewhere the origin is very far away from the triplet, and another where the origin is at the `target`.

[the remark this answers is not in the document]
Comment (bresch) [reply src/lib/motion_planning/PositionSmoothing.cpp:116]: @Luka-Filipovic You're right, this code assumes `a + b = c` in a triangle, which is only correct if the triangle is a line segment... Let me quickly fix this.

[the remark this answers is not in the document]
Comment (Luka-Filipovic) [reply src/lib/motion_planning/PositionSmoothing.cpp:116]: FWIW I will share my approach at achieving the same thing. I didn't figure out a way to do it properly using the `computeMaxSpeedInWaypoint`, there doesn't seem to be a correct way to define the "vertical angle" when all 3 points can be anywhere in 3d space.

So instead, I calculated the speed constraints when the `target` is reached and the `next` becomes a new target. These xy and z speeds are then clamped to each other's norms. 
```
	vel_sp_at_target // unit vector, target -> next, any large magnitude
	xy_speed_at_target // computeMaxSpeedFromDistance() from target to next
	z_speed_at_target // computeMaxSpeedFromDistance() from target to next

	math::trajectory::clampToXYNorm(vel_sp_at_target, xy_speed_at_target, 0.5f);
	math::trajectory::clampToZNorm(vel_sp_at_target, z_speed_at_target, 0.5f);

	float xy_speed_limit_at_target = vel_sp_at_target.xy().norm();
	float z_speed_limit_at_target = fabsf(vel_sp_at_target.z());

	// check if triplet changes vertical direction at the target waypoint
	if ((target.z > prev.z && target.z > next.z) || (target.z < prev.z && target.z < next.z)) {
 		arrival_z_speed = 0.;
	} else {
		arrival_z_speed = z_speed_limit_at_target;
	}
```

[the remark this answers is not in the document]
Comment (bresch) [reply src/lib/motion_planning/PositionSmoothing.cpp:116]: Would that work: https://github.com/PX4/PX4-Autopilot/pull/24227 ?
